pkg/gofr/datasource/opentsdb/annotation.go

```diff
@@ -0,0 +1,329 @@
+	TSUIDs []string `json:"tsuids,omitempty"`
+
+	// StartTime is the Unix epoch timestamp for the start of the deletion request.
+	StartTime int64 `json:"startTime,omitempty"`
+
+	// EndTime is the optional Unix epoch timestamp for when the events ended.
+	EndTime int64 `json:"endTime,omitempty"`
+
+	// Global indicates whether global annotations should be deleted for the given time range.
+	Global bool `json:"global,omitempty"`
+}
+
+// BulkDeleteResp contains the results of a bulk annotation delete operation.
+type BulkDeleteResp struct {
+	BulkAnnotationDeleteInfo
+}
+
+func (bulkAnnotResp *BulkAnnotationResponse) SetStatus(code int) {
+	setStatus(bulkAnnotResp.ctx, bulkAnnotResp, code, "SetStatus-BulkAnnotation", bulkAnnotResp.logger)
+}
+
+func (bulkAnnotResp *BulkAnnotationResponse) setStatusCode(code int) {
+	bulkAnnotResp.StatusCode = code
+}
+
+func (bulkAnnotResp *BulkAnnotationResponse) GetCustomParser() func(respCnt []byte) error {
+	return getCustomParser(bulkAnnotResp.ctx, bulkAnnotResp, "GetCustomParser-BulkAnnotation", bulkAnnotResp.logger,
+		func(resp []byte) error {
+			originContents := string(resp)
+			tag := strings.Split(originContents, " ")[0]
+
+			var resultBytes []byte
+
+			if strings.Contains(tag, "error") || strings.Contains(tag, "totalDeleted") {
+				resultBytes = resp
+			} else if strings.Contains(tag, "tsuid") {
+				resultBytes = []byte(fmt.Sprintf(`{"InvolvedAnnotations":%s}`, originContents))
+			} else {
+				return fmt.Errorf("unrecognized bulk annotation response: %s", originContents)
+			}
+
+			return json.Unmarshal(resultBytes, &bulkAnnotResp)
+		})
+}
+
+func (bulkAnnotResp *BulkAnnotationResponse) String() string {
+	return toString(bulkAnnotResp.ctx, bulkAnnotResp, "ToString-BulkAnnotation", bulkAnnotResp.logger)
+}
+
+func (c *Client) BulkUpdateAnnotations(annotations []Annotation) (*BulkAnnotationResponse, error) {
+	span := c.addTrace(c.ctx, "BulkUpdateAnnotations")
+
+	status := StatusFailed
+
+	var message string
+
+	defer sendOperationStats(c.logger, time.Now(), "BulkUpdateAnnotations", &status, &message, span)
+
+	if len(annotations) == 0 {
+		message = "The annotations list is empty."
+		return nil, errors.New(message)
+	}
+
+	bulkAnnoEndpoint := fmt.Sprintf("%s%s", c.tsdbEndpoint, BulkAnnotationPath)
+
+	reqBodyCnt, err := marshalAnnotations(annotations)
+	if err != nil {
+		message = fmt.Sprintf("Failed to marshal annotations: %v", err)
+		return nil, errors.New(message)
+	}
+
+	bulkAnnoResp := BulkAnnotationResponse{logger: c.logger, tracer: c.tracer, ctx: c.ctx}
+	if err = c.sendRequest(http.MethodPost, bulkAnnoEndpoint, reqBodyCnt, &bulkAnnoResp); err != nil {
+		message = fmt.Sprintf("error while processing update bulk annotations request to url %q: %s", bulkAnnoEndpoint, err)
+		return nil, err
+	}
+
+	status = StatusSuccess
+	message = fmt.Sprintf("Bulk annotation updation processed successfully at url %q", bulkAnnoEndpoint)
+
+	c.logger.Logf("Bulk annotation updated successfully")
```

Comment:
Why not adding Log to the Logger interface?

I would use this here

```suggestion
	c.logger.Log("Bulk annotation updated successfully")
```